frontend/test/pytest/test_jvpvjp.py

```diff
@@ -0,0 +1,169 @@
+    return (
+        sum(((i if isinstance(i, tuple) else (i,)) for i in x), ()) if isinstance(x, tuple) else x
+    )
+
+
+testvec = [
+    (
+        lambda x1, x2: (
+            jnp.stack(
+                [
+                    3 * jnp.reshape(x1, [6]) + 4 * jnp.reshape(x2, [6]),
+                    3 * jnp.reshape(x1, [6]) + 4 * jnp.reshape(x2, [6]),
+                ]
+            )
+        ),
+        [jnp.zeros([3, 2], dtype=float), jnp.zeros([2, 3], dtype=float)],
+        [jnp.ones([3, 2], dtype=float), jnp.ones([2, 3], dtype=float)],
+        [jnp.ones([2, 6], dtype=float)],
+    ),
+    (
+        lambda x: (x, jnp.stack([1 * x, 2 * x, 3 * x])),
+        [jnp.zeros([4], dtype=float)],
+        [jnp.ones([4], dtype=float)],
+        [jnp.ones([4], dtype=float), jnp.ones([3, 4], dtype=float)],
+    ),
+    (
+        lambda x: jnp.stack([1 * x, 2 * x, 3 * x]),
+        [jnp.zeros([4], dtype=float)],
+        [jnp.ones([4], dtype=float)],
+        [jnp.ones([3, 4], dtype=float)],
+    ),
+    (
+        lambda x1, x2: (
+            1 * jnp.reshape(x1, [6]) + 2 * jnp.reshape(x2, [6]),
+            jnp.stack(
+                [
+                    3 * jnp.reshape(x1, [6]) + 4 * jnp.reshape(x2, [6]),
+                    3 * jnp.reshape(x1, [6]) + 4 * jnp.reshape(x2, [6]),
+                ]
+            ),
+        ),
+        [jnp.zeros([3, 2], dtype=float), jnp.zeros([2, 3], dtype=float)],
+        [jnp.ones([3, 2], dtype=float), jnp.ones([2, 3], dtype=float)],
+        [jnp.ones([6], dtype=float), jnp.ones([2, 6], dtype=float)],
+    ),
+]
+
+
+@pytest.mark.parametrize("f, x, t, _", testvec)
+def test_jvp_against_jax(f: callable, x: list, t: list, _):
+    """Numerically tests Catalyst's jvp against the JAX version."""
+
+    @qjit
+    def C_workflow():
+        return C_jvp(f, x, t, method="fd", argnum=list(range(len(x))))
+
+    @jax.jit
+    def J_workflow():
+        y, ft = J_jvp(f, *x)
+        return flatten_if_tuples((y, ft(*t)))
+
+    r1 = C_workflow()
+    r2 = J_workflow()
+    print(r1)
+    print(r2)
+
+    for a, b in zip(r1, r2):
+        assert_allclose(a, b, rtol=1e-6, atol=1e-6)
+
+
+@pytest.mark.parametrize("f, x, _, ct", testvec)
+def test_vjp_against_jax(f: callable, x: list, _, ct: list):
+    """Numerically tests Catalyst's jvp against the JAX version."""
+
+    @qjit
+    def C_workflow():
+        return C_vjp(f, x, ct, method="fd", argnum=list(range(len(x))))
+
+    @jax.jit
+    def J_workflow():
+        y, ft = J_vjp(f, *x)
+        ct2 = tree_unflatten(tree_flatten(y)[1], ct)
+        return flatten_if_tuples((y, ft(ct2)))
+
+    r1 = C_workflow()
+    r2 = J_workflow()
+    print(r1)
+    print(r2)
+
+    for a, b in zip(r1, r2):
+        assert_allclose(a, b, rtol=1e-6, atol=1e-6)
+
+
+@pytest.mark.parametrize("f, x, t, ct", testvec[0:1])
+def test_jvpvjp_argcheck(f: callable, x: list, t: list, ct: list):
+    """Numerically tests Catalyst's jvp against the JAX version."""
```

Comment:
My comment here: I think that they support classes in pytest to let users put some common initialization in constructors. I believe we do not need this feature now (do not have anything to initialize), that is why the current approach.

Reply:
From having worked with the PL test suite and debugging tests, I do heavily prefer test functions that are "as dumb as possible". Essentially it should be easy to isolate a test case without having to stitch various pieces of code together from different locations in the test file. I believe the core team has also adopted this strategy.

A few things to keep in mind when writing tests:
- each piece of logic being tested should live in a separate test function
- only parametrize tests if the same logic is tested with different input values
- prefer self-contained test functions with duplicated code if necessary (within reason)
- avoid test function with too much logic, that is overly modular / parametrized, or makes heavy use of fixtures and other setup code

Reply:
💯 we've gone through multiple styles in PL proper, and what @dime10 is the approach we have settled on. It can be tempting to write super clever tests that parametrize logic, but we found that tests that were 'dumb' -- even if code was duplicated! -- resulted in much more maintainable tests long term.

Reply:
I made this test take no arguments. Let me know if you think that other tests also look to complicated.

Reply:
I also add comments to the items of the testvector data, to highlight its purpose.

Reply:
@josh146 @dime10 I have modified the tests to follow these guidelines. At the same time, I must admit that I am not convinced even after some discussions that we have, and still think that these approach leads to disadvantage for the current project at least. But I obviously can not question your past experience, so I think that the right thing from my side would be to ask you just to make this points explicit by moving them from this comment to some project document like contribution guide, development guide or like this. What do you think, is it possible?

Reply:
I asked the core team if they had a written guide for PL tests, unfortunately it doesn't appear so. There is some guidelines in our [internal reviewer guide](https://github.com/PennyLaneAI/guidance-docs/blob/master/development/code-review.md#code-author-and-reviewers-checklist) that echo the above comments though.

Reply:
> I asked the core team if they had a written guide for PL tests, unfortunately it doesn't appear so. There is some guidelines in our [internal reviewer guide](https://github.com/PennyLaneAI/guidance-docs/blob/master/development/code-review.md#code-author-and-reviewers-checklist) that echo the above comments though.

Thank you, I got it!